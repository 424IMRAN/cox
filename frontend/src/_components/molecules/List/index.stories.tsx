import { Story, Meta } from '@storybook/react'
import { ThemeProvider } from '@material-ui/core/styles'
import MoreHorizIcon from '@material-ui/icons/MoreHoriz'
import List from '.'
import baseTheme from '../../../core-utils/theme'

export default {
  title: 'molecules/ListField',
  component: List,
  argTypes: {},
} as Meta

const Template: Story = ({ list, ...args }) => (
  <ThemeProvider theme={baseTheme}>
    <List list={list} {...args} />
  </ThemeProvider>
)

export const CheckedList = Template.bind({})
CheckedList.args = {
  list: [
    { checked: true, value: 'hello' },
    { checked: false, value: 'test' },
    { checked: true, value: 'world' },
  ],
  checkedList: true,
}

export const BasicList = Template.bind({})
BasicList.args = {
  list: [{ value: 'hello' }, { value: 'test' }, { value: 'world' }],
}

export const ListWithRightIcon = Template.bind({})
ListWithRightIcon.args = {
  list: [
    { checked: true, value: 'hello' },
    { checked: false, value: 'test' },
    { checked: true, value: 'world' },
  ],
  rightIcon: <MoreHorizIcon />,
}
