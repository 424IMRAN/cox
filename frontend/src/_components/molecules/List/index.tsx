/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import MuiList from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction'
import ListItemText from '@material-ui/core/ListItemText'
import clsx from 'clsx'
import Checkbox from '../../atoms/CheckBox'
import Typography from '../../atoms/Typography'
import { COLORS } from '../../../core-utils/theme'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
      maxWidth: 360,
      backgroundColor: theme.palette.background.paper,
    },
    primaryText: {
      color: COLORS.TEXT_500,
    },
    checkBoxStyle: {
      paddingRight: '12px',
    },
    defaultItemStyle: {
      paddingBottom: '4px',
      paddingTop: '4px',
    },
  })
)

type ListProps = {
  list: Array<any>
  rightIcon?: React.ReactNode
  checkedList?: boolean
  onClick?: (index: number | string, value: any) => void
  handleActionClick?: any
  customStyle?: any
  itemStyle?: any
  secondaryItemStyle?: any
}

const List = ({
  list,
  checkedList,
  rightIcon,
  onClick,
  handleActionClick,
  customStyle,
  itemStyle,
  secondaryItemStyle,
}: ListProps) => {
  const classes = useStyles()

  const handleClick = (value: number | string, idx: number) => () => {
    onClick?.(idx, value)
  }

  return (
    <MuiList className={clsx(classes.root, customStyle)}>
      {list.map((item, idx) => {
        return (
          <ListItem
            key={item.value}
            button
            onClick={handleClick(item?.value, idx)}
            classes={{ root: clsx(classes.defaultItemStyle, itemStyle) }}
          >
            {checkedList && (
              <ListItemIcon>
                <Checkbox
                  isChecked={item.checked}
                  tabIndex={-1}
                  disableRipple
                  customStyle={classes.checkBoxStyle}
                />
              </ListItemIcon>
            )}

            <ListItemText
              disableTypography
              primary={
                <Typography variant="body2" className={classes.primaryText}>
                  {item.value}
                </Typography>
              }
            />
            <ListItemSecondaryAction
              className={secondaryItemStyle}
              onClick={handleActionClick}
            >
              {rightIcon}
            </ListItemSecondaryAction>
          </ListItem>
        )
      })}
    </MuiList>
  )
}

List.defaultProps = {
  checkedList: false,
  customStyle: null,
  rightIcon: null,
  onClick: null,
  itemStyle: null,
  secondaryItemStyle: null,
  handleActionClick: null,
}

export default List
