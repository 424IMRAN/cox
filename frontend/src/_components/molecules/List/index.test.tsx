import React from 'react'
import ListItem from '@material-ui/core/ListItem'
import { mount, shallow } from 'enzyme'
import List from '.'

describe('List  Component', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<List list={[{ value: 'hello' }]} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Allows us to set props', () => {
    const wrapper = mount(<List list={[{ value: 'hello' }]} checkedList />)
    expect(wrapper.props().checkedList).toEqual(true)
    expect(wrapper.props().list).toEqual([{ value: 'hello' }])
  })

  it('onClick works', () => {
    const mockHandleChangeText = jest.fn()
    const component = mount(
      <List
        list={[{ value: 'hello' }]}
        checkedList
        onClick={mockHandleChangeText}
      />
    )
    component
      ?.find(ListItem)
      ?.props()
      ?.onClick?.({ target: {} } as any)
    expect(mockHandleChangeText).toHaveBeenCalledWith(0, 'hello')
  })
})
