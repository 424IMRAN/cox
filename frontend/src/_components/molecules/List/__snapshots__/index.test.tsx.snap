// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List  Component Renders correctly 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    key="hello"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemText))
      disableTypography={true}
      primary={
        <Typography
          align="left"
          className="makeStyles-primaryText-2"
          variant="body2"
        >
          hello
        </Typography>
      }
    />
    <WithStyles(ForwardRef(ListItemSecondaryAction)) />
  </WithStyles(ForwardRef(ListItem))>
</WithStyles(ForwardRef(List))>
`;
